### PR TITLE
[Quantization] fix attention quantization of gpt_oss model

### PR DESCRIPTION
**Pull request by xuebwang-amd**
## Purpose
There are mainly two aiming purposes of this PR:

1. Enable attention quantization (FP8, MXFP4, UnquantizedLinearMethod, _etc_) by passing `quant_config` to `OAIAttention` and associated calling classes.
2. Change the [packed_modules_mapping](https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/gpt_oss.py#L643) from `{"qkv": ["q_proj", "k_proj", "v_proj"]}` to `{"qkv_proj": ["q_proj", "k_proj", "v_proj"]}`. Here, the latter one `{"qkv_proj": ["q_proj", "k_proj", "v_proj"]}` is the common naming by most LLMs such as:
  - Llama: https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/llama.py#L512
  - Qwen3: https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/qwen3.py#L260
  - Mixtral: https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/mixtral.py#L498
This is essential because the layer name is named with suffix [qkv_proj](https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/gpt_oss.py#L99). Without this change, afterwards quantization pipeline could break, such as in AMD-Quark quantization:

`(Pdb) proj_name
'qkv_proj'
(Pdb) fused_mapping
{'qkv': ['q_proj', 'k_proj', 'v_proj']}`

The dictionary parsing https://github.com/vllm-project/vllm/blob/3a4255c7c4b73cd87b02129acf876168af8d9651/vllm/model_executor/layers/quantization/quark/utils.py#L40 will fail.

## Test Plan
Coming soon.

### CODE REVIEW

File: vllm/model_executor/layers/quantization/quark/utils.py [hunk 1]
```diff
@@ -29,15 +29,19 @@ def should_ignore_layer(
...
+    
     proj_name = layer_name.split(".")[-1]
 
     # Fused layers like gate_up_proj or qkv_proj will not be fused
     # in the safetensors checkpoint. So, we convert the name
     # from the fused version to unfused + check to make sure that
     # each shard of the fused layer has the same scheme.
-    if proj_name in fused_mapping:
+    if any(name_pattern in proj_name for name_pattern in fused_mapping.keys()): # proj_name in fused_mapping:
         shard_proj_names = fused_mapping[proj_name]
```
[the remark this answers is not in the document]
Comment (xuebwang-amd) [reply vllm/model_executor/layers/quantization/quark/utils.py:41]: I have made this change (commit [9aac383e4](https://github.com/vllm-project/vllm/pull/27334/commits/9aac383e46a7025a75f62a43a86ec42b8e27ab8e)) reverted since it's just a workaround.
Essentially, change the [packed_modules_mapping](https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/gpt_oss.py#L643) from `{"qkv": ["q_proj", "k_proj", "v_proj"]}` to `{"qkv_proj": ["q_proj", "k_proj", "v_proj"]}` can fix this issue. Here, the latter one `{"qkv_proj": ["q_proj", "k_proj", "v_proj"]}` is the common naming by most LLMs such as:

- Llama: https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/llama.py#L512
- Qwen3: https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/qwen3.py#L260
- Mixtral: https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/mixtral.py#L498

File: vllm/model_executor/models/gpt_oss.py [hunk 2]
```diff
@@ -199,7 +199,7 @@ def __init__(
 
         self.layer_idx = extract_layer_index(prefix)
         self.attn = OAIAttention(
-            config, prefix=f"{prefix}.attn", cache_config=cache_config
+            config, prefix=f"{prefix}.attn", quant_config=quant_config, cache_config=cache_config
```
[the remark this answers is not in the document]
Comment (xuebwang-amd) [reply vllm/model_executor/models/gpt_oss.py]: No worry, it is very easy-to-see. I had already updated the all related code snippets.

File: vllm/model_executor/layers/quantization/quark/utils.py [hunk 3]
```diff
@@ -29,15 +29,19 @@ def should_ignore_layer(
     if layer_name is None:
         return False
 
-    # layer_name = model.layers.0.self_attn.qkv_proj
-    # proj_name = qkv_proj
+    abbreviated_module_name_mappings = {"attn": "self_attn"}
+
+    for abbreviated_module_name in abbreviated_module_name_mappings:
+        if abbreviated_module_name in layer_name:
+            layer_name = layer_name.replace(abbreviated_module_name, abbreviated_module_name_mappings[abbreviated_module_name])
```
[the remark this answers is not in the document]
Comment (xuebwang-amd) [reply vllm/model_executor/layers/quantization/quark/utils.py]: These lines are workarounds, I have made this change (commit [9aac383e4](https://github.com/vllm-project/vllm/pull/27334/commits/9aac383e46a7025a75f62a43a86ec42b8e27ab8e)) reverted.

File: vllm/model_executor/models/gpt_oss.py [hunk 4]
```diff
@@ -199,7 +199,7 @@ def __init__(
...
         )
```
[the remark this answers is not in the document]
Comment (xuebwang-amd) [reply vllm/model_executor/models/gpt_oss.py:215]: Yeah, it is useful, a good reminder.
As commented above https://github.com/vllm-project/vllm/pull/27334#discussion_r2454866773, it is very easy-to-see. I had just updated the all related code snippets separately.

File: vllm/model_executor/models/gpt_oss.py [hunk 5]
```diff
@@ -640,7 +643,7 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
 
 
 class GptOssForCausalLM(nn.Module, SupportsPP, SupportsEagle3, SupportsLoRA):
-    packed_modules_mapping = {"qkv": ["q_proj", "k_proj", "v_proj"]}
+    packed_modules_mapping = {"qkv_proj": ["q_proj", "k_proj", "v_proj"]}
```
Comment (BowenBao) [inline vllm/model_executor/models/gpt_oss.py:654]: Is it correct to understand the previous "qkv" naming is both wrong and unused? wrong due to mismatch with [postfix definition as you pointed in the existing code](https://github.com/vllm-project/vllm/blob/61089465a6101790635ed96c26df3e9a57d8d2c9/vllm/model_executor/models/gpt_oss.py#L99), and unused because no quantization being involved at this point?

**Review by BowenBao — APPROVED** [review]
Reasonable fix, LGTM

File: vllm/model_executor/models/gpt_oss.py (hunk 5 above)
Comment (SageMoore) [reply vllm/model_executor/models/gpt_oss.py:654]: I have the same question. Was this just broken before?

Comment (xuebwang-amd) [reply vllm/model_executor/models/gpt_oss.py:654]: @BowenBao is right, there were no broken before because no trigger of attention quantization.
After passing `quant_config `to `OAIAttention` (1st change of this PR), the issue would subsequently happen (that's the motivation of the 2nd change of this PR). @SageMoore

**Comment by xuebwang-amd** [comment]
Should we always include an e2e model test case at this stage? @SageMoore @gshtras @BowenBao

**Comment by xuebwang-amd** [comment]
> @xuebwang-amd if you have such a test model handy, one balanced approach could be to add a test case with model under amd/org, albeit private at first and test skipped in CI. I feel PR can be landed while model being cleared from legal, no need to block on that.

That's an excellent suggestion, @BowenBao, and the model's publicity certainly doesn't prevent us from pursuing it. However, for public models (sometimes even private), lots of additional effort is typically required for quantization accuracy. Since the primary goal of this PR is to establish the quantization flow itself, this could divert us from our immediate objective.
An alternative approach would be to integrate Quark's "online" quantization directly into the test script. However, this would likely introduce a long dependency chain and add considerable testing overhead.
Therefore, I'd like to proceed with your first option, with regardless of quantization accuracy at this moment.

**Review by SageMoore — APPROVED** [review]

**Review by gshtras — APPROVED** [review]
